src/Data/Set.purs

```diff
@@ -82,6 +83,31 @@ instance monoidSet :: Ord a => Monoid (Set a) where
 instance semigroupSet :: Ord a => Semigroup (Set a) where
   append = union
 
+-- | Sets of elements of a given monoid can be made into a semiring by taking
+-- | set union as the addition operation and set-wise multiplication (using the
+-- | monoid operation of the elements) as the multiplication operation. The
+-- | zero element is then the empty set, and the one element is the singleton
+-- | set containing the `mempty` element of the given monoid.  For example:
+-- |
+-- | ```
+-- | > zero :: Set String
+-- | (fromFoldable Nil)
+-- |
+-- | > one :: Set String
+-- | (fromFoldable ("" : Nil))
+-- |
+-- | > Set.fromFoldable ["a", "b"] + Set.fromFoldable ["c", "d", "e"]
+-- | (fromFoldable ("a" : "b" : "c" : "d" : "e" : Nil))
+-- |
+-- | > Set.fromFoldable ["a", "b"] * Set.fromFoldable ["c", "d", "e"]
+-- | (fromFoldable ("ac" : "ad" : "ae" : "bc" : "bd" : "be" : Nil))
+-- | ```
+instance semiringSet :: (Ord a, Monoid a) => Semiring (Set a) where
+  zero = empty
+  add = union
+  one = singleton mempty
```

Comment:
So my new intuition for this is that we're adding the elements of `a` as additively idempotent elements to a new `Semring` of terms of the form `a_1 + ... + a_n`, where `<>` induces the multiplicative structure. `singleton` is the embedding, and there is a retraction whenever `a` is itself a semiring.

So the idempotent bit is admittedly a bit of a weird restriction, but this intuition at least makes this less mysterious to me. I think we could drop idempotence if we were working with bags instead of sets.

Reply:
Yes, this is another of the examples from https://en.wikipedia.org/wiki/Semiring#Specific_examples:

> Similarly, if (M, e, ⋅) is a monoid, then the set of finite multisets in M forms a semiring. That is, an element is a pair (n, f), where n ∈ N is a natural number and f : {1, 2, ..., n} → M is a function. The additive unit is (0,!), where ! : ∅ → M is the unique function. The multiplicative unit is (1, {e}). The sum is given by (m,f) + (n,g) = (m + n, (f,g)) and the product is given by (m,f)⋅(n,g) = (mn, f⋅g). Similarly, the set of arbitrary multisets in M forms a complete semiring.

I'm now wondering if this semiring actually is the free semiring, though, or whether it's really the above example. The semiring in this PR definitely doesn't fit with the intuition of satisfying only the relations it's required to by the axioms.